mlflow/transformers.py

```diff
@@ -0,0 +1,793 @@
+    """
+    Performs inference of the task type, used in generating a pipeline object based on the
+    underlying model's intended use case. This utility relies on the definitions within the
+    transformers pipeline construction utility functions.
+
+    :param model: Either the model or the Pipeline object that the task will be extracted or
+                  inferred from
+    :return: The task type string
+    """
+    from transformers import PreTrainedModel, TFPreTrainedModel, FlaxPreTrainedModel, Pipeline
+    from transformers.pipelines import get_task
+
+    if isinstance(model, Pipeline):
+        return model.task
+    elif isinstance(model, (PreTrainedModel, TFPreTrainedModel, FlaxPreTrainedModel)):
+        # transformers provides guard conditions for potentially invalid entries
+        return get_task(model.name_or_path)
+    else:
+        raise MlflowException(
+            f"The provided model type: {type(model)} is not supported. "
+            "Supported model types are: PreTrainedModel or Pipeline."
+        )
+
+
+def _validate_transformers_task_type(task: str) -> None:
+    """
+    Validates that a given ``task`` type is supported by the ``transformers`` library.
+    """
+    from transformers.pipelines import get_supported_tasks
+
+    valid_tasks = get_supported_tasks()
+
+    if task not in valid_tasks:
+        raise MlflowException(
+            f"The task provided is invalid. '{task}' is not a supported task. "
+            f"Must be one of: {valid_tasks}"
+        )
+
+
+def _get_engine_type(model):
+    """
+    Determines the underlying execution engine for the model based on the 3 currently supported
+    deep learning framework backends: ``tensorflow``, ``torch``, or ``flax``.
+    """
+    from transformers import PreTrainedModel, TFPreTrainedModel, FlaxPreTrainedModel
+
+    for cls in model.__class__.__mro__:
+        if issubclass(cls, TFPreTrainedModel):
+            return "tensorflow"
+        elif issubclass(cls, PreTrainedModel):
+            return "torch"
+        elif issubclass(cls, FlaxPreTrainedModel):
+            return "flax"
+
+
+def _get_base_model_architecture(model_or_pipeline):
+    """
+    Extracts the base model architecture type from a submitted model.
+    """
+    from transformers import Pipeline
+
+    if isinstance(model_or_pipeline, Pipeline):
+        return model_or_pipeline.model.name_or_path
+    else:
+        return model_or_pipeline.name_or_path
+
+
+def _get_instance_type(model, base: bool):
+    """
+    Utility for extracting the saved object type or, if the `base` argument is set to `True`,
+    the base ABC type of the model.
+    """
+    if base:
+        from transformers import PreTrainedModel, TFPreTrainedModel, FlaxPreTrainedModel, Pipeline
+
+        def _get_model_base_class(model):
+            pipeline_abc = Pipeline
+            tf_abc = TFPreTrainedModel
+            torch_abc = PreTrainedModel
+            flax_abc = FlaxPreTrainedModel
```

Comment:
Do we need these variables? I think directly using the classes is more straightforward.

Reply:
good call. I was going to use those variables for another purpose but abstracted that away. Removed the variables :)